lymph/unilateral.py

```diff
@@ -708,21 +662,142 @@ def modalities(self, modality_spsn: Dict[Any, List[float]]):
+        that can detect microscopic metastases
 
-        if len(self.allowed_lnl_states) == 2:
-            self.binary_modality(modality_spsn)
-        if len(self.allowed_lnl_states) == 3:
-            self.trinary_modality(modality_spsn)
+        Args:
+            modality_spsn (Dict[Any, List[float]]):  dictionary with modality name and [sp,sn] list
+
+        Returns:
+            Dict[str,np.ndarray]: returns a dictionary with modality name and pathological confusion matrices
+        """
+        modality_dictionary = {}
+        for mod, spsn in modality_spsn.items():
+            try:
+                self._check_modality(mod,spsn)
+            except TypeError:
+                raise
+            except ValueError:
+                raise
+
+            sp, sn = spsn
+            modality_dictionary[mod] = np.array([[sp, 1. - sp],
+                                               [1. - sn, sn],
+                                               [1. - sn, sn]])
+        return modality_dictionary
+
+
+    def confusion_matrix(self,modality: str, matrix: np.ndarray):
+        """checks a manually inserted confusion matrix.
+
+        Args:
+            modality (str): name of the modality
+            matrix (np.ndarray): confusion matrix with number of allowed lnl states x 2 dimensionality
+
+        Raises:
+            ValueError: returns a type error if the modality is not a string
+            ValueError: raises a value error if the spec or sens is not a number btw. 0.5 and 1.0
+
+        Returns:
+            Dict[str, np.ndarray]: returns a dictionary with modality as name and the accepted confusion matrix
+        """
+        if matrix.shape != (len(self.allowed_lnl_states), 2):
+            msg = (f'the shape of the confusion matrix does not match the model '
+                   f'insert a ({len(self.allowed_lnl_states)}x2) matrix ')
+            raise ValueError(msg)
+        elif not (np.all(0 <= matrix) and np.all(matrix <= 1)):
+            msg = ("All values need to be between 0 and 1")
+            raise ValueError(msg)
+        else:
+            return {modality: matrix}
+
+
+    def binary_modality(self, modality_spsn: Dict[Any, List[float]]):
+        """Produces the confusion matrix for a binary model.
+
+        Args:
+            modality_spsn (Dict[Any, List[float]]): dictionary with modality name and [sp,sn] list
+
+        Returns:
+            Dict[str,np.ndarray]: returns a dictionary with modality name and confusion matrices
+        """
+        modality_dictionary = {}
+        for mod, spsn in modality_spsn.items():
+            try:
+                self._check_modality(mod,spsn)
+            except TypeError:
+                raise
+            except ValueError:
+                raise
+
+            sp, sn = spsn
+            modality_dictionary[mod] = np.array([[sp     , 1. - sp],
+                                               [1. - sn, sn     ]])
+        return modality_dictionary

```

Comment:
Maybe in the binary case, the `Unilateral` class can just take the 2x3 matrix and delete the middle column.

Reply:
in theory yes, but to do so a 2x3 matrix needs to be produced first. (Either clinical or pathological) which would mean, that we would also need to apply the clinical or pathological function, which doesn't make too much sense. 
Alternatively the `modalities` setter would not require dictionaries with confusion matrices, which would then again defeat the idea of using the functions `clinical` and `pathological`.
What could be done is to define the setter to handle dictionaries with spsn lists as a binary case, use the clinical or pathological function on it and then delete the middle column. (which would work, but I am not sure whether this is truly more elegant)

Reply:
Why doesn't using the `clinical` or `pathological` function in the binary case make sense?

The way I imagined it, the `modalities` setter expects a dictionary with modality names as keys and 2x3 confusion matrices as values in the trinary case. For that, the two helper functions would work like a charm while still allowing the user to define their own weird modality.

For the binary case, we can allow 2x2 _OR_ 2x3 matrices. In the latter case we can just delete the unnecessary values. That way, we can still use the `clinical` and the `pathological` function where we only pass sensitivity and specificity. Although there wouldn't be any intrinsic difference for those two in the binary case.